### DISCUSSION TITLE
How to Use This with CCXT?

### DISCUSSION THREAD

**Discussion by mutedstack** [bmoscon/cryptofeed#410] [General]
My program:

1. Check balance using ccxt REST
2. Get order books of two symbols from cryptofeed ws
3. Create two orders at same time using ccxt REST

How and when to get order books of two symbols from cryptofeed ws? Should I get order book when callback is called and wait until callbacks of both symbols are called or store order book to local and get from local cache? What is the frequency to create orders?

Problem of getting order book from local cache instead of from callback: Suppose server pushes order book every 20ms, so the local cache remains same in the 20ms, if you create order every 1ms according to this, you will create 20 orders. But order book should change after you create first order, so you create order according to the outdated order book.

How to handle exception in cryptofeed? Should I exit if can't get order book from cryptofeed?

@pottertech Do you any advice regarding you mentioned ccxt at https://github.com/bmoscon/cryptofeed/issues/163 ?

**Comment by bmoscon** [comment]
How you design your application is up to you. I would separate out data collection and order execution. Something is responsible for storing data and something else is responsible for using that data and verifying its correctness (i.e. do you have data for 2 symbols, how recent is the data, etc). I don't think CCXT has anything to do with this, its a question of how do you design an application.

**Reply by mutedstack** [reply]
Does the data collection service provide data to all the other services? If so, when you want to update data collection service, all the other services need to wait.

**Reply by bmoscon** [reply]
i dont know why that would be true, but as I said, its your system, so whatever design works for you

**Comment by yohplala** [comment]
Hi @Jay54520 
Are you really in the ms ?
What exchange are you targetting?
If I take Binance, they indicate that through WS, order book get refreshed every 100 or 1000ms...
https://github.com/binance/binance-spot-api-docs/blob/master/web-socket-streams.md#partial-book-depth-streams

**Comment by mutedstack** [comment]
I have tried only to return order book within X ms, following codes will wait until order book is not None, this can work but need to increase X for some less active markets.
